art/classifiers/classifier.py

```diff
@@ -57,35 +56,186 @@ def __init__(self, channel_index, clip_values=None, defences=None, preprocessing
+
+    def _apply_preprocessing(self, x, y, fit):
+        """
+        Apply all defences and preprocessing operations on the inputs `(x, y)`. This function has to be applied to all
+        raw inputs (x, y) provided to the classifier.
+
+        :param x: Features, where first dimension is the number of samples.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels), where first dimension is the number of samples.
+        :type y: `np.ndarray`
+        :param fit: `True` if the defences are applied during training.
+        :return: Value of the data after applying the defences.
+        :rtype: `np.ndarray`
+        """
+        x_preprocessed, y_preprocessed = self._apply_preprocessing_defences(x, y, fit=fit)
+        x_preprocessed = self._apply_preprocessing_standardisation(x_preprocessed)
+        return x_preprocessed, y_preprocessed
+
+    def _apply_preprocessing_defences(self, x, y, fit=False):
+        """
+        Apply all defences of the classifier on the raw inputs `(x, y)`. This function is intended to only be called
+        from function `_apply_defences_and_preprocessing`.
+
+        :param x: Features, where first dimension is the number of samples.
+        :type x: `np.ndarray`
+        :param y: Target values (class labels), where first dimension is the number of samples.
+        :type y: `np.ndarray`
+        :param fit: `True` if the function is call before fit/training and `False` if the function is called before a
+                    predict operation
+        :return: Arrays for `x` and `y` after applying the defences.
+        :rtype: `np.ndarray`
+        """
+        if self.defences is not None:
+            for defence in self.defences:
+                if fit:
+                    if defence.apply_fit:
+                        x, y = defence(x, y)
+                else:
+                    if defence.apply_predict:
+                        x, y = defence(x, y)
+
+        return x, y
+
+    def _apply_preprocessing_standardisation(self, x):
+        """
+        Apply standardisation to input data `x`.
+
+        :param x: Input data, where first dimension is the number of samples.
+        :type x: `np.ndarray`
+        :return: Array for `x` with the standardized data.
+        :rtype: `np.ndarray`
+        """
+        if self.preprocessing is not None:
+            sub, div = self.preprocessing
+            sub = np.asarray(sub, dtype=x.dtype)
+            div = np.asarray(div, dtype=x.dtype)
+
+            res = x - sub
+            res = res / div
+
+        else:
+            res = x
+
+        return res
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        attributes = {(k[1:], v) if k[0] == '_' else (k, v) for (k, v) in self.__dict__.items()}
+        attributes = ['{}={}'.format(k, v) for (k, v) in attributes]
+        repr_string = class_name + '(' + ', '.join(attributes) + ')'
+        return repr_string
+
+
+class ClassifierNeuralNetwork(ABC):
```

Comment:
I would vote to keep the three interfaces, because I can already recognise the need for fourth interface for tree-based classifiers in the near future. The `ClassifierNeuralNetwork` interface contains all the neural network specific functionalities like e.g. `learning_phase` or `get_activations` that are not needed in the any other interface. Only `Classifier` is a complete interface, whereas the other two interfaces can be mixed in with `Classifier`. That way `AdversarialPatch` requires both `Classifier` and `ClassifierNeuralNetwork` because it only makes sense for convolutional neural networks, as far as we know.